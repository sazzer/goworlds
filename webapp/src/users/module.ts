/** The module prefix */
export const MODULE_PREFIX = 'Users/';
