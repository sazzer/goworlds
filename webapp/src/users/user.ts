/** The shape of a user */
export declare type User = {
    id: string,
    created: string,
    updated: string,
    name: string,
    email: string,
};