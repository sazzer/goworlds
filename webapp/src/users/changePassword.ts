//////// Action for changing the users password

import {MODULE_PREFIX} from "./module";
import {
    Action,
    asyncAction,
    buildAction,
    buildSaga,
    failedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request} from "../api";
import {User} from "./user";

/** The action type for changing the users password */
export const CHANGE_PASSWORD_ACTION = MODULE_PREFIX + 'changePassword';

/** The shape of the action to change the users password */
declare type ChangePasswordAction = {
    userId: string,
    password: string,
    callback: (err: Error | undefined) => void
}

/**
 * Build the Redux action to change the users password
 * @param userId The ID of the User to update
 * @param password The new password
 * @param callback A callback to trigger when the action is completed
 */
export function changePassword(userId: string, password: string,
                                  callback?: (err: Error | undefined) => void): Action<ChangePasswordAction> {
    return buildAction(CHANGE_PASSWORD_ACTION, {
        userId,
        password,
        callback: callback || (() => {})
    });
}

/**
 * Saga to change a users password
 * @param action the action to handle
 */
export function* changePasswordSaga(action: Action<ChangePasswordAction>) : IterableIterator<any> {
    yield asyncAction(CHANGE_PASSWORD_ACTION, async () => {
        const response = await request<any>('/users/{id}', {
            method: 'PATCH',
            urlParams: {
                id: action.payload.userId
            },
            body: {
                password: action.payload.password,
            }
        });

        return {
            ...response.body,
        };
    }, action.payload);
}

/**
 * Saga for when changing a users password was a success
 * @param action the action to handle
 */
export function changePasswordSuccessSaga(action: ResolvedAsyncAction<ChangePasswordAction, User>) {
    if (action.input) {
        action.input.callback(undefined);
    }
}

/**
 * Saga for when changing a users password was a failure
 * @param action the action to handle
 */
export function changePasswordFailedSaga(action: ResolvedAsyncAction<ChangePasswordAction, Error>) {
    if (action.input) {
        action.input.callback(action.payload);
    }
}

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(CHANGE_PASSWORD_ACTION, changePasswordSaga),
    buildSaga(succeededAction(CHANGE_PASSWORD_ACTION), changePasswordSuccessSaga),
    buildSaga(failedAction(CHANGE_PASSWORD_ACTION), changePasswordFailedSaga),
];
