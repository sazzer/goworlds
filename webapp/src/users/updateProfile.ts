//////// Action for updating the user profile

import {MODULE_PREFIX} from "./module";
import {
    Action,
    asyncAction,
    buildAction,
    buildSaga,
    failedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request} from "../api";
import {User} from "./user";

/** The action type for updating the users profile */
export const UPDATE_PROFILE_ACTION = MODULE_PREFIX + 'updateProfile';

/** The shape of the action to update the users profile */
declare type UpdateUserProfileAction = {
    userId: string,
    email: string,
    name: string,
    callback: (err: Error | undefined) => void
}

/**
 * Build the Redux action to update the users profile
 * @param userId The ID of the User to update
 * @param email The new email address of the user
 * @param name The new name of the user
 * @param callback A callback to trigger when the action is completed
 */
export function updateUserProfile(userId: string, email: string, name: string,
                                  callback?: (err: Error | undefined) => void): Action<UpdateUserProfileAction> {
    return buildAction(UPDATE_PROFILE_ACTION, {
        userId,
        email,
        name,
        callback: callback || (() => {})
    });
}

/**
 * Saga to update a user profile
 * @param action the action to handle
 */
export function* updateUserProfileSaga(action: Action<UpdateUserProfileAction>) : IterableIterator<any> {
    yield asyncAction(UPDATE_PROFILE_ACTION, async () => {
        const response = await request<any>('/users/{id}', {
            method: 'PATCH',
            urlParams: {
                id: action.payload.userId
            },
            body: {
                name: action.payload.name,
                email: action.payload.email,
            }
        });

        return {
            ...response.body,
        };
    }, action.payload);
}

/**
 * Saga for when updating a user profile was a success
 * @param action the action to handle
 */
export function updateUserProfileSuccessSaga(action: ResolvedAsyncAction<UpdateUserProfileAction, User>) {
    if (action.input) {
        action.input.callback(undefined);
    }
}

/**
 * Saga for when updating a user profile was a failure
 * @param action the action to handle
 */
export function updateUserProfileFailedSaga(action: ResolvedAsyncAction<UpdateUserProfileAction, Error>) {
    if (action.input) {
        action.input.callback(action.payload);
    }
}

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(UPDATE_PROFILE_ACTION, updateUserProfileSaga),
    buildSaga(succeededAction(UPDATE_PROFILE_ACTION), updateUserProfileSuccessSaga),
    buildSaga(failedAction(UPDATE_PROFILE_ACTION), updateUserProfileFailedSaga),
];
