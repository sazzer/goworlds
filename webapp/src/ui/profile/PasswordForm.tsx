import React, {FunctionComponent, useState} from 'react';
import {Button, Form, Message} from "semantic-ui-react";
import {useTranslation} from "react-i18next";
import * as Yup from "yup";
import {StringSchema} from "yup";
import {Formik} from "formik";
import {FormikErrorMessage} from "../common/FormikErrorMessage";
import {ErrorMessage} from "../common/ErrorMessage";
import {User} from "../../users/user";
import {ProblemError} from "../../api";
import {useDispatch} from "react-redux";
import {changePassword} from "../../users/changePassword";

/** The props that the PasswordForm area needs */
type PasswordFormProps = {
    user: User,
};

/**
 * Component for changing the password for a user
 * @constructor
 */
export const PasswordForm: FunctionComponent<PasswordFormProps> = ({user}) => {
    const { t, i18n } = useTranslation();
    const dispatch = useDispatch();
    const [submitting, setSubmitting] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);
    const [success, setSuccess] = useState<boolean>(false);

    const schema = Yup.object().shape({
        password: Yup.string()
            .required(t('profile.password.password.errors.required')),
        password2: Yup.string()
            .required(t('profile.password.password2.errors.required'))
            .when('password', (password: string, schema: StringSchema) =>
                schema.oneOf([password], t('profile.password.password2.errors.match'))),
    });

    const doSubmit = (password: string) => {
        setSubmitting(true);
        setError(undefined);
        setSuccess(false);

        dispatch(changePassword(user.id, password, (err) => {
            setSubmitting(false);
            if (err) {
                if (err instanceof ProblemError) {
                    if (i18n.exists('profile.password.submit.errors.' + err.type)) {
                        setError(err.type);
                    } else {
                        setError('unexpected_error');
                    }
                } else {
                    setError('unexpected_error');
                }
            } else {
                setSuccess(true);
            }
        }));
    };

    return (
        <Formik initialValues={{password: '', password2: ''}}
                validationSchema={schema}
                onSubmit={(values) => doSubmit(values.password)}>
            {({values, isValid, errors, touched, handleSubmit, handleChange, handleBlur, handleReset, dirty}) =>
                <Form onSubmit={handleSubmit}
                      error={!isValid || error !== undefined}
                      loading={submitting}
                      data-test="PasswordForm">
                    <Form.Field required data-test="password">
                        <label>
                            {t('profile.password.password.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('profile.password.password.placeholder')}
                                    name="password"
                                    type="password"
                                    value={values.password}
                                    error={touched.password && errors.password !== undefined}
                                    onChange={handleChange}
                                    onBlur={handleBlur}
                                    autoFocus />
                        <FormikErrorMessage name="password" />
                    </Form.Field>
                    <Form.Field required data-test="password2">
                        <label>
                            {t('profile.password.password2.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('profile.password.password2.placeholder')}
                                    name="password2"
                                    type="password"
                                    value={values.password2}
                                    error={touched.password2 && errors.password2 !== undefined}
                                    onChange={handleChange}
                                    onBlur={handleBlur} />
                        <FormikErrorMessage name="password2" />
                    </Form.Field>
                    <Button type="submit" primary disabled={!dirty}>
                        {t('profile.password.submit.save')}
                    </Button>
                    <ErrorMessage testName="PasswordFormErrors" errors={[
                        error && t('profile.password.submit.errors.' + error)
                    ]}/>
                    { success && <Message positive>{t('profile.password.submit.success')}</Message> }
                </Form>
            }
        </Formik>
    );
};
