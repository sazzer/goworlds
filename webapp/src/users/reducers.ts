import {reducers as users} from "./users";

export default {
    users,
};
