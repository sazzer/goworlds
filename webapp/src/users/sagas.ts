import { sagas as usersSagas } from './users';
import { sagas as updateProfileSagas } from './updateProfile';
import { sagas as changePasswordSagas } from './changePassword';

export default [
    ...usersSagas,
    ...updateProfileSagas,
    ...changePasswordSagas,
];
