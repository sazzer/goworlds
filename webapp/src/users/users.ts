import {
    Action, AsyncAction,
    asyncAction,
    buildAction,
    buildSaga,
    buildSelector, finishedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request} from "../api";
import {MODULE_PREFIX} from "./module";
import {createReducer} from "redux-create-reducer";
import produce from "immer";
import {select} from 'redux-saga/effects';
import {UPDATE_PROFILE_ACTION} from "./updateProfile";
import {User} from "./user";

/** The shape of the state for this module */
export declare type State = {
    users: Array<User>,
};

/** The initial state for this module */
const INITIAL_STATE: State = {
    users: [],
};

//////// Action for loading a user

/** The action type for registering a new user */
const LOAD_USER_ACTION = MODULE_PREFIX + 'loadUser';

/** The shape of the action to load a User by ID */
declare type LoadUserAction = {
    userId: string,
    force: boolean,
    callback: () => void
}

/**
 * Build the Redux action to load a user
 * @param userId The ID of the User to load
 * @param force Set to true to re-load the user even if we've already got a copy of it
 * @param callback Callback to trigger on loading the user
 */
export function loadUser(userId: string, force: boolean = false, callback?: () => void): Action<LoadUserAction> {
    return buildAction(LOAD_USER_ACTION, {
        userId,
        force,
        callback: callback || (() => {}),
    });
}

/**
 * Saga to load a user
 * @param action the action to handle
 */
export function* loadUserSaga(action: Action<LoadUserAction>) : IterableIterator<any> {
    let doAction = action.payload.force;
    if (!doAction) {
        const existingUser = yield select(selectUserById(action.payload.userId));
        doAction = existingUser === undefined;
    }

    if (doAction) {
        yield asyncAction(LOAD_USER_ACTION, async () => {
            const response = await request<any>('/users/{id}', {
                method: 'GET',
                urlParams: {
                    id: action.payload.userId
                }
            });

            return {
                ...response.body,
            };
        }, action.payload);
    }
}

/**
 * Trigger the callback when loading the user has finished
 * @param action the action to handle
 */
export function loadUserFinishedSaga(action: AsyncAction<LoadUserAction>) {
    if (action.input) {
        action.input.callback();
    }
}

/**
 * Reducer for storing a user
 * @param state the state to update
 * @param action the action to update the state with
 * @return the new state
 */
export function storeUserReducer(state: State, action: ResolvedAsyncAction<LoadUserAction, User>) : State {
    return produce(state, (draft: State) => {
        if (action.input === undefined || action.payload === undefined) {
            return;
        }

        const id = action.payload.id;
        const users = draft.users.filter(user => user.id !== id);

        draft.users = [
            action.payload,
            ...users,
        ];
    });
}

//////// Selectors for getting user data

/** Selector to get the specified user */
export const selectUserById = (id: string) =>
    buildSelector(['users', 'users'], (users: Array<User>) =>
        users.find(user => user.id === id));

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(LOAD_USER_ACTION, loadUserSaga),
    buildSaga(finishedAction(LOAD_USER_ACTION), loadUserFinishedSaga),
];

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [succeededAction(LOAD_USER_ACTION)] : storeUserReducer,
    [succeededAction(UPDATE_PROFILE_ACTION)] : storeUserReducer,
});
