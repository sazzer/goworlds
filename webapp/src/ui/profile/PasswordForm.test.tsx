import React from 'react';
import {PasswordForm} from './PasswordForm';
import {fireEvent, render} from '@testing-library/react';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {User} from "../../users/user";

/** Set up the component to test */
function setup(user: User) {
    const mockStoreCreator = configureStore();
    const store = mockStoreCreator({});

    const element = render(<Provider store={store}><PasswordForm user={user} /></Provider>);

    const enterPassword = (value: string) => {
        fireEvent.change(element.getByPlaceholderText('Enter Password'), {
            target: {
                value,
            }
        });
    };

    const reenterPassword = (value: string) => {
        fireEvent.change(element.getByPlaceholderText('Re-enter Password'), {
            target: {
                value,
            }
        });
    };

    const submitForm = () => new Promise(resolve => {
        fireEvent.click(element.getByText('Change Password'));
        setImmediate(resolve);
    });

    return {
        store,

        element,

        enterPassword,
        reenterPassword,
        submitForm,
    };
}

it('renders the initial state when a user is present', () => {
    const {element} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    expect(element.baseElement).toMatchSnapshot();
});

it('renders correctly after entering new matching details', () => {
    const {element, enterPassword, reenterPassword} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterPassword('newPassword');
    reenterPassword('newPassword');

    expect(element.baseElement).toMatchSnapshot();
});

it('renders correctly after entering new mismatched details', () => {
    const {element, enterPassword, reenterPassword} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterPassword('newPassword');
    reenterPassword('different');

    expect(element.baseElement).toMatchSnapshot();
});

it('renders errors if submitting when blank', async () => {
    const {element, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Marks as loading when submitting valid details', async () => {
    const {element, enterPassword, reenterPassword, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterPassword('newPassword');
    reenterPassword('newPassword');
    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Doesn\'t mark as loading when submitting mismatched details', async () => {
    const {element, enterPassword, reenterPassword, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterPassword('newPassword');
    reenterPassword('different');
    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Triggers the correct redux action when submitting valid details', async () => {
    const {store, enterPassword, reenterPassword, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    store.clearActions();

    enterPassword('newPassword');
    reenterPassword('newPassword');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()[0]).toMatchObject({
        type: 'Users/changePassword',
        payload: {
            userId: 'abc123',
            password: 'newPassword'
        }
    });
});

it('Doesn\'t trigger any redux action when submitting mismatched details', async () => {
    const {store, enterPassword, reenterPassword, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    store.clearActions();

    enterPassword('newPassword');
    reenterPassword('different');
    await submitForm();

    expect(store.getActions()).toHaveLength(0);
});

it('Renders correctly when submitting triggers the callback as success', async () => {
    const {element, store, enterPassword, reenterPassword, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    store.clearActions();

    enterPassword('newPassword');
    reenterPassword('newPassword');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(undefined);

    expect(element.baseElement).toMatchSnapshot();
});

it('Renders correctly when submitting triggers the callback as an unexpected failure', async () => {
    const {element, store, enterPassword, reenterPassword, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    store.clearActions();

    enterPassword('newPassword');
    reenterPassword('newPassword');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(new Error());

    expect(element.baseElement).toMatchSnapshot();
});