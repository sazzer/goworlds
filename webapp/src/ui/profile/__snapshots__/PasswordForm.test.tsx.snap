// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doesn't mark as loading when submitting mismatched details 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="error field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value="different"
            />
          </div>
        </div>
        <div
          class="ui error message"
        >
          Passwords do not match
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Marks as loading when submitting valid details 1`] = `
<body>
  <div>
    <form
      class="ui loading form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Renders correctly when submitting triggers the callback as an unexpected failure 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      >
        !!profile.password.submit.errors.unexpected_error!!
      </div>
    </form>
  </div>
</body>
`;

exports[`Renders correctly when submitting triggers the callback as success 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
      <div
        class="ui positive message"
      >
        Password Changed
      </div>
    </form>
  </div>
</body>
`;

exports[`renders correctly after entering new matching details 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`renders correctly after entering new mismatched details 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="newPassword"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value="different"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`renders errors if submitting when blank 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary disabled button"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`renders the initial state when a user is present 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="PasswordForm"
    >
      <div
        class="required field"
        data-test="password"
      >
        <label>
          New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password2"
      >
        <label>
          Re-enter New Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password2"
              placeholder="Re-enter Password"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary disabled button"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        Change Password
      </button>
      <div
        class="ui error message"
        data-test="PasswordFormErrors"
      />
    </form>
  </div>
</body>
`;
